# list of software
## iSim
https://github.com/csbl/iSim
Rawls, Kristopher D., et al. "A simplified metabolic network reconstruction to promote understanding and development of flux balance analysis tools." Computers in Biology and Medicine 105 (2019): 64-71. 

## transcript2flux
https://github.com/cdanielmachado/transcript2flux
D. Machado, M.J. Herrgard. Systematic Evaluation of Methods for Integration of Transcriptomic Data into Constraint-Based Models of Metabolism. PLoS Computational Biology, 10(4), 2014.

- implement iMat and Lee2012

## Lee2012
https://link.springer.com/article/10.1186/1752-0509-6-73
Lee, Dave, et al. "Improving metabolic flux predictions using absolute gene expression data." BMC systems biology 6 (2012): 1-9.

Input data and matlab code could be accessed in [Additional information](https://link.springer.com/article/10.1186/1752-0509-6-73#additional-information).

I put it in `Lee_BMC-Sys-Bio-2012/original_Lee2012_code`

## gembox
https://github.com/ruppinlab/gembox

- In-house toolbox for genome-scale metabolic modeling (GEM) for the Ruppin lab. This is an R package that implements GEM functionalities.

Showcase: https://github.com/ruppinlab/covid_metabolism
Cheng, Kuoyuan, et al. "Genome‐scale metabolic modeling reveals SARS‐CoV‐2‐induced metabolic changes and antiviral targets." Molecular systems biology 17.11 (2021): e10260.

## MERGE
https://github.com/WalhoutLab/MERGE
Yilmaz, Lutfu Safak, et al. "Modeling tissue‐relevant Caenorhabditis elegans metabolism at network, pathway, reaction, and metabolite levels." Molecular Systems Biology 16.10 (2020): e9649.

## DeltaFBA
https://github.com/CABSEL/DeltaFBA
Ravi, Sudharshan, and Rudiyanto Gunawan. "ΔFBA—Predicting metabolic flux alterations using genome-scale metabolic models and differential transcriptomic data." PLoS Computational Biology 17.11 (2021): e1009589.

## FluxSamplingComparison
https://github.com/HAHerrmann/FluxSamplingComparison
Herrmann, Helena A., et al. "Flux sampling is a powerful tool to study metabolism under changing environmental conditions." NPJ systems biology and applications 5.1 (2019): 32.

## GEM Models
H. Wang, J. L. Robinson, P. Kocabas, J. Gustafsson, M. Anton, P.-E. Cholley, et al. Genome-scale metabolic network reconstruction of model animals as a platform for translational research. PNAS 118, e2102344118 (2021).

### Human
https://github.com/SysBioChalmers/Human-GEM
online info: https://sysbiochalmers.github.io/Human-GEM-guide/getting_started/

### Fruit fly
https://github.com/SysBioChalmers/Fruitfly-GEM

### Yeast 
YeastsModels-master.zip
https://github.com/SysBioChalmers/YeastsModels/tree/master

yeast-GEM-main.zip
https://github.com/SysBioChalmers/yeast-GEM

